The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10410162179/job/28831366881) [(plain)](https://github.com/rust-lang/rust/commit/ac8a698713965896012f3a92bb092754e0b9f9ac/checks/28831366881/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 2.707 Building wheels for collected packages: reuse
#16 2.708   Building wheel for reuse (pyproject.toml): started
#16 2.954   Building wheel for reuse (pyproject.toml): finished with status 'done'
#16 2.956   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#16 2.956   Stored in directory: /tmp/pip-ephem-wheel-cache-006dbptb/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#16 2.958 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#16 3.357 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#16 3.357 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#16 DONE 3.4s
---
configure: run `python /checkout/x.py --help`
##[endgroup]
Attempting with retry: make prepare
downloading https://static.rust-lang.org/dist/2024-07-26/rust-std-beta-x86_64-unknown-linux-gnu.tar.xz
curl 7.81.0 (x86_64-pc-linux-gnu) libcurl/7.81.0 OpenSSL/3.0.2 zlib/1.2.11 brotli/1.0.9 zstd/1.4.8 libidn2/2.3.2 libpsl/0.21.0 (+libidn2/2.3.2) libssh/0.9.6/openssl/zlib nghttp2/1.43.0 librtmp/2.3 OpenLDAP/2.5.18
Release-Date: 2022-01-05
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtmp rtsp scp sftp smb smbs smtp smtps telnet tftp 
Features: alt-svc AsynchDNS brotli GSS-API HSTS HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz NTLM NTLM_WB PSL SPNEGO SSL TLS-SRP UnixSockets zstd
  File "/checkout/src/bootstrap/bootstrap.py", line 1212, in <module>
    main()
  File "/checkout/src/bootstrap/bootstrap.py", line 1197, in main
    bootstrap(args)
    bootstrap(args)
  File "/checkout/src/bootstrap/bootstrap.py", line 1162, in bootstrap
    build.download_toolchain()
  File "/checkout/src/bootstrap/bootstrap.py", line 599, in download_toolchain
    download_component(download_info)
  File "/checkout/src/bootstrap/bootstrap.py", line 462, in download_component
    get(
  File "/checkout/src/bootstrap/bootstrap.py", line 70, in get
    download(temp_path, "{}/{}".format(base, url), True, verbose)
  File "/checkout/src/bootstrap/bootstrap.py", line 88, in download
    _download(path, url, probably_big, verbose, True)
  File "/checkout/src/bootstrap/bootstrap.py", line 113, in _download
    if curl_version() > 7.70:
  File "/checkout/src/bootstrap/bootstrap.py", line 83, in curl_version
    return re.match("^curl ([0-9]+\\.[0-9]+)", run(["curl", "-V"]))[1]
  File "/usr/lib/python3.10/re.py", line 190, in match
    return _compile(pattern, flags).match(string)
TypeError: expected string or bytes-like object
Command failed. Attempt 2/5:
downloading https://static.rust-lang.org/dist/2024-07-26/rust-std-beta-x86_64-unknown-linux-gnu.tar.xz
downloading https://static.rust-lang.org/dist/2024-07-26/rust-std-beta-x86_64-unknown-linux-gnu.tar.xz
curl 7.81.0 (x86_64-pc-linux-gnu) libcurl/7.81.0 OpenSSL/3.0.2 zlib/1.2.11 brotli/1.0.9 zstd/1.4.8 libidn2/2.3.2 libpsl/0.21.0 (+libidn2/2.3.2) libssh/0.9.6/openssl/zlib nghttp2/1.43.0 librtmp/2.3 OpenLDAP/2.5.18
Release-Date: 2022-01-05
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtmp rtsp scp sftp smb smbs smtp smtps telnet tftp 
Features: alt-svc AsynchDNS brotli GSS-API HSTS HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz NTLM NTLM_WB PSL SPNEGO SSL TLS-SRP UnixSockets zstd
  File "/checkout/src/bootstrap/bootstrap.py", line 1212, in <module>
    main()
  File "/checkout/src/bootstrap/bootstrap.py", line 1197, in main
    bootstrap(args)
    bootstrap(args)
  File "/checkout/src/bootstrap/bootstrap.py", line 1162, in bootstrap
    build.download_toolchain()
  File "/checkout/src/bootstrap/bootstrap.py", line 599, in download_toolchain
    download_component(download_info)
  File "/checkout/src/bootstrap/bootstrap.py", line 462, in download_component
    get(
  File "/checkout/src/bootstrap/bootstrap.py", line 70, in get
    download(temp_path, "{}/{}".format(base, url), True, verbose)
  File "/checkout/src/bootstrap/bootstrap.py", line 88, in download
    _download(path, url, probably_big, verbose, True)
  File "/checkout/src/bootstrap/bootstrap.py", line 113, in _download
    if curl_version() > 7.70:
  File "/checkout/src/bootstrap/bootstrap.py", line 83, in curl_version
    return re.match("^curl ([0-9]+\\.[0-9]+)", run(["curl", "-V"]))[1]
  File "/usr/lib/python3.10/re.py", line 190, in match
    return _compile(pattern, flags).match(string)
TypeError: expected string or bytes-like object
Command failed. Attempt 3/5:
downloading https://static.rust-lang.org/dist/2024-07-26/rust-std-beta-x86_64-unknown-linux-gnu.tar.xz
downloading https://static.rust-lang.org/dist/2024-07-26/rust-std-beta-x86_64-unknown-linux-gnu.tar.xz
curl 7.81.0 (x86_64-pc-linux-gnu) libcurl/7.81.0 OpenSSL/3.0.2 zlib/1.2.11 brotli/1.0.9 zstd/1.4.8 libidn2/2.3.2 libpsl/0.21.0 (+libidn2/2.3.2) libssh/0.9.6/openssl/zlib nghttp2/1.43.0 librtmp/2.3 OpenLDAP/2.5.18
Release-Date: 2022-01-05
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtmp rtsp scp sftp smb smbs smtp smtps telnet tftp 
Features: alt-svc AsynchDNS brotli GSS-API HSTS HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz NTLM NTLM_WB PSL SPNEGO SSL TLS-SRP UnixSockets zstd
  File "/checkout/src/bootstrap/bootstrap.py", line 1212, in <module>
    main()
  File "/checkout/src/bootstrap/bootstrap.py", line 1197, in main
    bootstrap(args)
    bootstrap(args)
  File "/checkout/src/bootstrap/bootstrap.py", line 1162, in bootstrap
    build.download_toolchain()
  File "/checkout/src/bootstrap/bootstrap.py", line 599, in download_toolchain
    download_component(download_info)
  File "/checkout/src/bootstrap/bootstrap.py", line 462, in download_component
    get(
  File "/checkout/src/bootstrap/bootstrap.py", line 70, in get
    download(temp_path, "{}/{}".format(base, url), True, verbose)
  File "/checkout/src/bootstrap/bootstrap.py", line 88, in download
    _download(path, url, probably_big, verbose, True)
  File "/checkout/src/bootstrap/bootstrap.py", line 113, in _download
    if curl_version() > 7.70:
  File "/checkout/src/bootstrap/bootstrap.py", line 83, in curl_version
    return re.match("^curl ([0-9]+\\.[0-9]+)", run(["curl", "-V"]))[1]
  File "/usr/lib/python3.10/re.py", line 190, in match
    return _compile(pattern, flags).match(string)
TypeError: expected string or bytes-like object
Command failed. Attempt 4/5:
downloading https://static.rust-lang.org/dist/2024-07-26/rust-std-beta-x86_64-unknown-linux-gnu.tar.xz
downloading https://static.rust-lang.org/dist/2024-07-26/rust-std-beta-x86_64-unknown-linux-gnu.tar.xz
curl 7.81.0 (x86_64-pc-linux-gnu) libcurl/7.81.0 OpenSSL/3.0.2 zlib/1.2.11 brotli/1.0.9 zstd/1.4.8 libidn2/2.3.2 libpsl/0.21.0 (+libidn2/2.3.2) libssh/0.9.6/openssl/zlib nghttp2/1.43.0 librtmp/2.3 OpenLDAP/2.5.18
Release-Date: 2022-01-05
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtmp rtsp scp sftp smb smbs smtp smtps telnet tftp 
Features: alt-svc AsynchDNS brotli GSS-API HSTS HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz NTLM NTLM_WB PSL SPNEGO SSL TLS-SRP UnixSockets zstd
  File "/checkout/src/bootstrap/bootstrap.py", line 1212, in <module>
    main()
  File "/checkout/src/bootstrap/bootstrap.py", line 1197, in main
    bootstrap(args)
    bootstrap(args)
  File "/checkout/src/bootstrap/bootstrap.py", line 1162, in bootstrap
    build.download_toolchain()
  File "/checkout/src/bootstrap/bootstrap.py", line 599, in download_toolchain
    download_component(download_info)
  File "/checkout/src/bootstrap/bootstrap.py", line 462, in download_component
    get(
  File "/checkout/src/bootstrap/bootstrap.py", line 70, in get
    download(temp_path, "{}/{}".format(base, url), True, verbose)
  File "/checkout/src/bootstrap/bootstrap.py", line 88, in download
    _download(path, url, probably_big, verbose, True)
  File "/checkout/src/bootstrap/bootstrap.py", line 113, in _download
    if curl_version() > 7.70:
  File "/checkout/src/bootstrap/bootstrap.py", line 83, in curl_version
    return re.match("^curl ([0-9]+\\.[0-9]+)", run(["curl", "-V"]))[1]
  File "/usr/lib/python3.10/re.py", line 190, in match
    return _compile(pattern, flags).match(string)
TypeError: expected string or bytes-like object
Command failed. Attempt 5/5:
downloading https://static.rust-lang.org/dist/2024-07-26/rust-std-beta-x86_64-unknown-linux-gnu.tar.xz
downloading https://static.rust-lang.org/dist/2024-07-26/rust-std-beta-x86_64-unknown-linux-gnu.tar.xz
curl 7.81.0 (x86_64-pc-linux-gnu) libcurl/7.81.0 OpenSSL/3.0.2 zlib/1.2.11 brotli/1.0.9 zstd/1.4.8 libidn2/2.3.2 libpsl/0.21.0 (+libidn2/2.3.2) libssh/0.9.6/openssl/zlib nghttp2/1.43.0 librtmp/2.3 OpenLDAP/2.5.18
Release-Date: 2022-01-05
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtmp rtsp scp sftp smb smbs smtp smtps telnet tftp 
Features: alt-svc AsynchDNS brotli GSS-API HSTS HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz NTLM NTLM_WB PSL SPNEGO SSL TLS-SRP UnixSockets zstd
  File "/checkout/src/bootstrap/bootstrap.py", line 1212, in <module>
    main()
  File "/checkout/src/bootstrap/bootstrap.py", line 1197, in main
    bootstrap(args)
    bootstrap(args)
  File "/checkout/src/bootstrap/bootstrap.py", line 1162, in bootstrap
    build.download_toolchain()
  File "/checkout/src/bootstrap/bootstrap.py", line 599, in download_toolchain
    download_component(download_info)
  File "/checkout/src/bootstrap/bootstrap.py", line 462, in download_component
    get(
  File "/checkout/src/bootstrap/bootstrap.py", line 70, in get
    download(temp_path, "{}/{}".format(base, url), True, verbose)
  File "/checkout/src/bootstrap/bootstrap.py", line 88, in download
    _download(path, url, probably_big, verbose, True)
  File "/checkout/src/bootstrap/bootstrap.py", line 113, in _download
    if curl_version() > 7.70:
  File "/checkout/src/bootstrap/bootstrap.py", line 83, in curl_version
    return re.match("^curl ([0-9]+\\.[0-9]+)", run(["curl", "-V"]))[1]
  File "/usr/lib/python3.10/re.py", line 190, in match
    return _compile(pattern, flags).match(string)
TypeError: expected string or bytes-like object
The command has failed after 5 attempts.
  local time: Thu Aug 15 20:37:18 UTC 2024
  network time: Thu, 15 Aug 2024 20:37:18 GMT
##[error]Process completed with exit code 1.
```

</details>

--- a/src/bootstrap/bootstrap.py
+++ b/src/bootstrap/bootstrap.py
@@ -79,6 +79,8 @@ def get(base, url, path, checksums, verbose=False):
                 eprint("removing", temp_path)
             os.unlink(temp_path)
 
+def curl_version():
+    return float(re.match(bytes("^curl ([0-9]+\\.[0-9]+)", "utf8"), require(["curl", "-V"]))[1])
 
 def download(path, url, probably_big, verbose):
     for _ in range(4):
@@ -107,11 +109,15 @@ def _download(path, url, probably_big, verbose, exception):
         # If curl is not present on Win32, we should not sys.exit
         #   but raise `CalledProcessError` or `OSError` instead
         require(["curl", "--version"], exception=platform_is_win32())
-        run(["curl", option,
+        extra_flags = []
+        if curl_version() > 7.70:
+            extra_flags = [ "--retry-all-errors" ]
+        run(["curl", option] + extra_flags + [
             "-L", # Follow redirect.
             "-y", "30", "-Y", "10",    # timeout if speed is < 10 bytes/sec for > 30 seconds
             "--connect-timeout", "30",  # timeout if cannot connect within 30 seconds
             "-o", path,
+            "--continue-at", "-",
             "--retry", "3", "-SRf", url],
             verbose=verbose,
             exception=True, # Will raise RuntimeError on failure

--- a/src/bootstrap/src/core/download.rs
+++ b/src/bootstrap/src/core/download.rs
@@ -26,16 +26,8 @@ fn extract_curl_version(out: &[u8]) -> f32 {
     let Some(i) = out.iter().position(|&x| x == b' ') else { return 0.0 };
     let out = &out[..i];
     let Some(k) = out.iter().rev().position(|&x| x == b'.') else { return 0.0 };
-    let out = &out[..out.len()-k-1];
+    let out = &out[..out.len() - k - 1];
     std::str::from_utf8(out).unwrap().parse().unwrap_or(0.0)
-}
-
-#[test]
-fn test_extract_curl_version() {
-    assert_eq!(extract_curl_version(b"\
-        curl 8.4.0 (x86_64-pc-linux-gnu) libcurl/8.4.0 \
-        OpenSSL/3.0.13 zlib/1.3 brotli/1.1.0 zstd/1.5.5 libidn2/2.3.4 \
-        libssh2/1.11.0 nghttp2/1.57.0"), 8.4);
 }
 
 fn curl_version() -> f32 {

--- a/src/bootstrap/src/core/download.rs
+++ b/src/bootstrap/src/core/download.rs
@@ -21,6 +21,31 @@ fn try_run(config: &Config, cmd: &mut Command) -> Result<(), ()> {
     config.try_run(cmd)
 }
 
+fn extract_curl_version(out: &[u8]) -> f32 {
+    let out = &out[5..];
+    let Some(i) = out.iter().position(|&x| x == b' ') else { return 0.0 };
+    let out = &out[..i];
+    let Some(k) = out.iter().rev().position(|&x| x == b'.') else { return 0.0 };
+    let out = &out[..out.len()-k-1];
+    std::str::from_utf8(out).unwrap().parse().unwrap_or(0.0)
+}
+
+#[test]
+fn test_extract_curl_version() {
+    assert_eq!(extract_curl_version(b"\
+        curl 8.4.0 (x86_64-pc-linux-gnu) libcurl/8.4.0 \
+        OpenSSL/3.0.13 zlib/1.3 brotli/1.1.0 zstd/1.5.5 libidn2/2.3.4 \
+        libssh2/1.11.0 nghttp2/1.57.0"), 8.4);
+}
+
+fn curl_version() -> f32 {
+    let mut curl = Command::new("curl");
+    curl.arg("-V");
+    let Ok(out) = curl.output() else { return 0.0 };
+    let out = out.stdout;
+    extract_curl_version(&out)
+}
+
 /// Generic helpers that are useful anywhere in bootstrap.
 impl Config {
     pub fn is_verbose(&self) -> bool {
@@ -219,6 +244,8 @@ impl Config {
             "30", // timeout if cannot connect within 30 seconds
             "-o",
             tempfile.to_str().unwrap(),
+            "--continue-at",
+            "-",
             "--retry",
             "3",
             "-SRf",
@@ -228,6 +255,10 @@ impl Config {
             curl.arg("-s");
         } else {
             curl.arg("--progress-bar");
+        }
+        // --retry-all-errors was added in 7.71.0, don't use it if curl is old.
+        if dbg!(curl_version()) > 7.70 {
+            curl.arg("--retry-all-errors");
         }
         curl.arg(url);
         if !self.check_run(&mut curl) {

